assets/mdk/brittleness.mdk

```diff
@@ -0,0 +1,179 @@
+
+<!-- inline-dafny RationalAdd/M1 -->
+
+This function will verify without too much difficulty using Dafny 3.13.1, but not at all with Dafny 4.x. Even with Dafny 3.13.1, verification is brittle. Removing `ensures Rational(add(x, y))` leads to a failure to prove the other `ensures` clause, even though `assert Rational(r)` is a statement of the same fact. Adding `var r1 := x + y` also leads to the postcondition not being verified, even though `r1` is never used (and even though `assert r == x + y` succeeds).
+
+# Measuring and reducing brittleness
+
+Note that the last of these commented lines, `assert r == x + y`, hints at one of the concepts we’ll use to help reduce brittleness, because brittleness tends to be closely related to the *difficulty* of verification. Execution time is one way to measure difficulty, but execution time can vary widely, especially when doing verification on cloud-hosted computers. Another measure of difficulty, with greater reproducibility, is the Z3-specific *resource count*. If you verify the example `add` function in VS Code (with the Dafny extension installed), you can hover over the function signature line and see a popup that lists “Total resource usage” in units labeled RU (for “resource units”). This number gives you an indication of how much work it took the SMT solver to verify a particular definition. And we’ll see that reducing this number helps reduce brittleness. In general, we tend to aim for less than around 200K RU per definition. For the `add` function we’re considering, using Dafny 3.13, the reported RU is around 679K on my Intel-based Mac laptop.
+
+Before getting into specifics for how to reduce resource use, let’s step back for a moment to discuss what makes automated proofs difficult. The process that an automated prover goes through to establish the validity of a particular statement can be thought of as searching through the space of possible proofs. Roughly speaking, the prover has a starting point (facts that it takes as true at the beginning, such as those in preconditions and assumptions) and an ending point (facts that it’s trying to prove, such as those in postconditions and assertions). To get from the starting point to the ending point, it can repeatedly take a step by applying any one of a set of possible rules. The number of facts available at the starting point is directly proportional to the number rules available to apply, and the prover’s job is to successfully choose the rule that will most quickly get it closer to the ending point. The best rule is difficult to predict, however, so the choice is sometimes (or frequently) arbitrary. This means that the most reliable ways to make the prover’s job easier are to 1) reduce the number of alternative steps available to choose from at any given point, and 2) reduce the total length of the sequence of steps that must be taken to get from the starting point to the ending point.
```

Comment:
arbitrary. Mentions that SMT provers are in part effective because of the heuristics they use?